### ISSUE TITLE
Simplify design of source and target

### ISSUE THREAD

**Issue by pmaria** [kg-construct/rml-target-source-spec#18]
In the current draft both the `Source` and `Target` involve properties to describe at least the following aspects of a resource:
- encoding
- compression

Additionally, a distinction is made between `Source`/`Target` and `Access`, which creates a rather complex chain of resources. And, when we look at how Access is applied in the current proposal, its intent is to model a Data Source.

I think there is room to simplify the current design, by removing the distinction between `Source`/`Target` and `Access`. And I think we can reuse the same `Source` class for both the source and target side.

[![](https://mermaid.ink/img/pako:eNpVkD1vgzAQhv8K8pxAgZYkDJ3arVs6ernYh7FkbGSfK6WI_15DPio83b1-7r2PiQknkbVMGAjhQ4PyMHDL7ZpnZxe9wGziNktPuGH0GIJ29iagTdXaqlsW0Gsw-hfoCeR5zu387_fllBZgtraa0AM5v5JbYr9_f8zQZmENFrM79A1eIW0hWjW2YwP6AbRMq61tOKMeB-SsTaHEDqIhzlLLhMZRAuGn1GkI1nZgAu4YRHLnqxWsJR_xAd0v9KSMA4mpaGJ0HZc7Kh2W7sLZTqtFj94kuScaQ1sUy3euNPXxkqdzFkHLHjz1P6emaKrmCFWNzaGGt7qW4lKejl31Wnby8FJWwOZ5_gMS9ZOB?type=png)](https://mermaid.live/edit#pako:eNpVkD1vgzAQhv8K8pxAgZYkDJ3arVs6ernYh7FkbGSfK6WI_15DPio83b1-7r2PiQknkbVMGAjhQ4PyMHDL7ZpnZxe9wGziNktPuGH0GIJ29iagTdXaqlsW0Gsw-hfoCeR5zu387_fllBZgtraa0AM5v5JbYr9_f8zQZmENFrM79A1eIW0hWjW2YwP6AbRMq61tOKMeB-SsTaHEDqIhzlLLhMZRAuGn1GkI1nZgAu4YRHLnqxWsJR_xAd0v9KSMA4mpaGJ0HZc7Kh2W7sLZTqtFj94kuScaQ1sUy3euNPXxkqdzFkHLHjz1P6emaKrmCFWNzaGGt7qW4lKejl31Wnby8FJWwOZ5_gMS9ZOB)

So in this way, we could have a mapping like this:

```turtle
<#DCATSourceAccess> a dcat:Dataset;
  dcat:distribution [ a dcat:Distribution;
    dcat:downloadURL "https://rml.io/specs/rml-target/Supergirl.xml";
  ];
.

<#TriplesMap> a rr:TriplesMap;
  rml:logicalSource [ a rml:LogicalSource;
    rml:source <#DCATSourceAccess> ;
    rml:referenceFormulation ql:JSONPath;
    rml:iterator "$.[*]";
  ];
  rr:subjectMap [ a rr:SubjectMap;
    rr:template "http://example.org/{@id}";
    rml:logicalTarget <#TargetDump1>;
  ];
  rr:predicateObjectMap [ a rr:PredicateObjectMap;
    rr:predicateMap [ a rr:PredicateMap;
      rr:constant foaf:name;
    ];
    rr:objectMap [ a rr:ObjectMap;
      rml:reference "name/text()";
      rml:logicalTarget <#TargetDump2>;
    ];
  ];
  rr:predicateObjectMap [ a rr:PredicateObjectMap;
    rr:predicateMap [ a rr:PredicateMap;
      rr:constant foaf:nickname;
    ];
    rr:objectMap [ a rr:ObjectMap;
      rml:reference "nickname/text()";
    ];
  ];
.

<#TargetDump1> a rmlt:LogicalTarget;
  rmlt:target <#VoIDDump1>;
.
<#TargetDump2> a rmlt:LogicalTarget;
  rmlt:target <#VoIDDump1>;
.

<#VoIDDump1> a void:Dataset ;
  void:dataDump <file:///data/dump1.nt>;
  void:feature formats:N-Triples;
  rmlt:serialization formats:N-Triples;
.
<#VoIDDump2> a void:Dataset ;
  void:dataDump <file:///data/dump2.nt>;
  void:feature formats:N-Triples;
  rmlt:serialization formats:N-Triples; 
.

```

A specific source type specification can then define how to infer source specific properties like encoding, compression, serialization/format, etc. Many standards already have properties that define these aspects. For example:
* [csvw:encoding](https://www.w3.org/ns/csvw)
* [void:feature](https://www.w3.org/TR/void/#features)
* [dcat:compressFormat](https://www.w3.org/TR/vocab-dcat-3/#Property:distribution_compression_format)
* [dcterms:format](https://www.w3.org/TR/vocab-dcat-3/#Property:distribution_format)

This would allow for more simply reusing already existing source descriptions as sources and targets, without having to duplicate information.

**Comment by DylanVanAssche** [comment]
This idea works for some vocabularies but does not cover the following cases:

- NULL values: how can we define which values should be considered NULL/empty?
- Queries: Some sources such as a SPARQL endpoint require a query to be supplied.
- Compression: Some sources may be compressed. The latter is not covered by all types of vocabulary such as CSVW. VoID can do it, but looses the MIME type then of the file itself as it would say `application/gzip` but nothing about the content of the file such as `text/csv; charset=UTF-8` 

These cases are the reasons why we have this access property to allow use to expand a Source or Target description without invading the external vocabularies.

**Comment by pmaria** [comment]
Right, I think we should look at this from an equivalence point of view. We want to be able to reuse existing descriptions of the things that we naturally see as being sources.

So, for example, we can state that `csvw:Table` is the equivalent of `rml:Source`, and we can specifying how the encoding of a `csvw:Table` can be mapped to `rml:encoding`. Yes, CSVW has a different, more nested way of specificying the encoding, but that is not a problem, as long as we know how to derive the one from the other.

For example in OWL we could state something like :
```ttl
rml:encoding owl:propertyChainAxiom ( csvw:dialect csvw:encoding ) .
```

From an implementation point of view, an engine could provide some interface methods with which the value for encoding, handling nulls, compression etc. can be derived from the particular source description. This should be described per source type.

In this way there is no invasion of external vocabularies.

**Comment by DylanVanAssche** [comment]
> Right, I think we should look at this from an equivalence point of view. We want to be able to reuse existing descriptions of the things that we naturally see as being sources.

My main concern is when the property is just absent in the vocabulary we re-use. For example: NULL values in a DCAT Distribution or the query to send to a SPARQL endpoint described with Service Description (SD) vocabulary. For those we would need to specify `rml:null` somewhere, the question is where mostly. The answer to that influences the necessity of `rml:access` or not.

TL;DR Should we have these properties (`rml:null`, `rml:encoding`, etc.) and where then if we drop `rml:access`?

> From an implementation point of view, an engine could provide some interface methods with which the value for encoding, handling nulls, compression etc. can be derived from the particular source description. This should be described per source type.

So pushing this problem to the engines? I cannot follow to be honest.

**Comment by pmaria** [comment]
> > Right, I think we should look at this from an equivalence point of view. We want to be able to reuse existing descriptions of the things that we naturally see as being sources.
> 
> My main concern is when the property is just absent in the vocabulary we re-use. For example: NULL values in a DCAT Distribution or the query to send to a SPARQL endpoint described with Service Description (SD) vocabulary. For those we would need to specify `rml:null` somewhere, the question is where mostly. The answer to that influences the necessity of `rml:access` or not.
> 
> TL;DR Should we have these properties (`rml:null`, `rml:encoding`, etc.) and where then if we drop `rml:access`?

Yes, we should. I would say on the Source object (`rml:Source`), as in the image above.

So e.g.

```ttl
<#TriplesMap> a rr:TriplesMap;
  rml:logicalSource [ a rml:LogicalSource;
    rml:source [
       csvw:dialect [
         csvw:encoding enc:UTF-8 ; 
       ] ;
       rml:encoding enc:UTF-8 ; # inferrable
       rml:compression comp:zip ;
       rml:null "" ;
    ] ;
    rml:referenceFormulation ql:JSONPath;
    rml:iterator "$.[*]";
  ];
```

Although you could argue that `rml:null` might be a better fit on the `rml:LogicalSource`, since it is not necessarily a property of the source itself, but specifies how queries or expressions on the source are to handle certain values.
It is also not something that you would use when specifying a target "source".

> > From an implementation point of view, an engine could provide some interface methods with which the value for encoding, handling nulls, compression etc. can be derived from the particular source description. This should be described per source type.
> 
> So pushing this problem to the engines? I cannot follow to be honest.

Just the problem of deriving the value for rml:encoding, rml:null etc. if specified in a specific source description like csvw, dcat etc.

**Comment by DylanVanAssche** [comment]
> Although you could argue that rml:null might be a better fit on the rml:LogicalSource, since it is not necessarily a property of the source itself, but specifies how queries or expressions on the source are to handle certain values.
It is also not something that you would use when specifying a target "source".

Problem with the example above is that the class is actually not a rml:Source:

```
<#TriplesMap> a rr:TriplesMap;
  rml:logicalSource [ a rml:LogicalSource;
    rml:source [ a csvw:Table # not a rml:Source
       csvw:dialect [
         csvw:encoding enc:UTF-8 ; 
       ] ;
       rml:encoding enc:UTF-8 ; # inferrable
       rml:compression comp:zip ;
       rml:null "" ;
    ] ;
    rml:referenceFormulation ql:JSONPath;
    rml:iterator "$.[*]";
  ];
```

So adding these properties would extend existing vocabulary each time which I tried to avoid with `rml:access`.
The only place would be in `rml:LogicalSource` then which is acceptable I think?
The `rml:source` property is then _only_ the access description, nothing more, as it was in the original RML spec.

> Just the problem of deriving the value for rml:encoding, rml:null etc. if specified in a specific source description like csvw, dcat etc.

Nothing needs to be inferred then, if the external vocabulary supports things like `csvw:encoding`, it is not required for the processor to infer this, but it could do that. I rather avoid it, but won't block it either.

**Comment by pmaria** [comment]
> Problem with the example above is that the class is actually not a rml:Source:
> 
> ```
> <#TriplesMap> a rr:TriplesMap;
>   rml:logicalSource [ a rml:LogicalSource;
>     rml:source [ a csvw:Table # not a rml:Source
>        csvw:dialect [
>          csvw:encoding enc:UTF-8 ; 
>        ] ;
>        rml:encoding enc:UTF-8 ; # inferrable
>        rml:compression comp:zip ;
>        rml:null "" ;
>     ] ;
>     rml:referenceFormulation ql:JSONPath;
>     rml:iterator "$.[*]";
>   ];
> ```
> 
> So adding these properties would extend existing vocabulary each time which I tried to avoid with `rml:access`. The only place would be in `rml:LogicalSource` then which is acceptable I think? The `rml:source` property is then _only_ the access description, nothing more, as it was in the original RML spec.
>

I disagree that we extend existing vocabulary. In the above example the CSVW vocabulary stays the same. We just semantically align it with `rml:Source`, saying in essence: things that are a `csvw:Table` can also be seen as an `rml:Source` and this is how a CSVW description can be translated to an RML description. To me this is exactly what RDF excels at and why we use it to make KGs.

**Comment by pmaria** [comment]
For extra clarification of what I mean, this would be the same example but with explicit typing of rml:Source as well:

```turtle
<#TriplesMap> a rr:TriplesMap;
  rml:logicalSource [ a rml:LogicalSource;
    rml:source [ a csvw:Table , rml:Source ;
       csvw:dialect [
         csvw:encoding enc:UTF-8 ; 
       ] ;
       rml:encoding enc:UTF-8 ; # inferrable
       rml:compression comp:zip ;
       rml:null "" ;
    ] ;
    rml:referenceFormulation ql:JSONPath;
    rml:iterator "$.[*]";
  ];
```

**Comment by DylanVanAssche** [comment]
We discussed this issue on the KGC Slack and came to the following consensus:

1. Drop `rml:access` & leave `rml:source` property open-ended, any IRI is acceptable. Via the `rdf:type` property, it is advertised whatever class is in there, thus engines can use this to determine if they support this source description or not.
2. We define our properties for `rml:Source` class like we do now for rml:null, etc. but these are not required as some vocabularies like CSVW have these properties builtin. Engines MUST support the ones from `rml:Source` and SHOULD support the ones from the vocabularies like CSVW (if they support the vocabulary of course).
3. In case of overlap between vocabularies and `rml:Source` and different values are given, e.g. `csvw:encoding` says UTF-16 while `rml:encoding` says UTF-8, `rml:Source` takes precedence.

**Comment by DylanVanAssche** [comment]
Implemented in f773d4b
